eth/downloader/downloader.go

```diff
@@ -1178,23 +1180,75 @@ func (d *Downloader) importBlockResults(results []*fetchResult) error {
 	for i, result := range results {
 		blocks[i] = types.NewBlockWithHeader(result.Header).WithBody(result.Transactions, result.Uncles, result.ExtTransactions, result.SubManifest)
 	}
-	if index, err := d.core.InsertChain(blocks); err != nil {
-		if err.Error() == core.ErrAddedFutureCache.Error() {
-			d.pauseCh<-true // Pause the downloader
-			return nil
-		}
-		if index < len(results) {
-			log.Debug("Downloaded item processing failed", "number", results[index].Header.Number(), "hash", results[index].Header.Hash(), "err", err)
-		} else {
-			// The InsertChain method in core.go will sometimes return an out-of-bounds index,
-			// when it needs to preprocess blocks to import a sidechain.
-			// The importer will put together a new list of blocks to import, which is a superset
-			// of the blocks delivered from the downloader, and the indexing will be off.
-			log.Debug("Downloaded item processing failed on sidechain import", "index", index, "err", err)
+	// Loop until all blocks have been inserted, or the process has been cancelled
+	count := 0
+	waitAppend := common.Hash{}
+	waitSubAppend := common.Hash{}
+	for {
+		select {
+		case <-d.quitCh:
```

Comment:
I observed that, after the sync, prime started another sync loop but, zone-0-2 did not and looking at the logs, it looks stuck. This leads me to think we didn't exit the downloader to start another sync. We might want to take a look at the cancel ch and use it to exit after sync. Or use both quit and cancel.

Reply:
Yeah, I think the importBlocks loop needs to listen to the cancelCh too